File: src/main/java/com/cutcalculator/ottimizzatore/MultiStartCasuale.java
package com.cutcalculator.ottimizzatore;

import com.cutcalculator.dominio.Avanzo;
import com.cutcalculator.formule.Distinta;

import java.util.List;
import java.util.Random;

/**
 * Meta-strategia multi-start, l'euristica <b>predefinita</b>: genera più piani candidati e tiene il
 * migliore — meno barre nuove, e a parità la {@link PianoDiTaglio#mediaGeometricaSfrido() media
 * geometrica degli sfridi} più bassa. È l'uso previsto di quella metrica.
 * <p>
 * Parte da un candidato deterministico (best-fit su ordine decrescente), poi prova
 * {@code iterazioni} impacchettamenti best-fit con i pezzi in ordine <b>casuale</b>
 * (stesso {@code seed} ⇒ risultato riproducibile). Best-fit su ordine perturbato a volte
 * batte l'ordine decrescente puro: tenendo il migliore non si può fare peggio del baseline.
 * <p>
 * Costa circa <b>4 volte</b> il best-fit da solo (misurato su 200 distinte da 20-100 pezzi: 0,7 ms
 * l'una), il che su ordini di questa taglia resta impercettibile. Se un domani le commesse
 * diventassero molto grandi, è la prima impostazione da riportare su "miglior incastro".
 */
public class MultiStartCasuale implements Ottimizzatore {

    private final int iterazioni;
    private final long seed;

    public MultiStartCasuale(int iterazioni, long seed) {
        this.iterazioni = iterazioni;
        this.seed = seed;
    }

    /** Default comodo: 50 tentativi, seed fisso per risultati riproducibili. */
    public MultiStartCasuale() {
        this(50, 0L);
    }

    @Override
    public PianoDiTaglio ottimizza(Distinta distinta, List<Avanzo> avanzi, double lunghezzaBarraStandard) {
        PianoDiTaglio migliore = Impacchettatore.impacchetta(distinta, avanzi, lunghezzaBarraStandard,
                Impacchettatore.DECRESCENTE, Impacchettatore.MIGLIOR_INCASTRO);

        Random rng = new Random(seed);
        for (int i = 0; i < iterazioni; i++) {
            PianoDiTaglio candidato = Impacchettatore.impacchetta(distinta, avanzi, lunghezzaBarraStandard,
                    Impacchettatore.casuale(rng), Impacchettatore.MIGLIOR_INCASTRO);
            if (meglio(candidato, migliore)) {
                migliore = candidato;
            }
        }
        return migliore;
    }

    /**
     * Il confronto fra due piani, in <b>quest'ordine</b>: prima chi compra <b>meno barre nuove</b>,
     * poi — a parità — chi ha la media geometrica dello sfrido più bassa.
     * <p>
     * La media geometrica da sola non basta come criterio: premia lo sfrido <i>concentrato</i>, e un
     * piano con una barra in più può avere una media migliore di uno con una barra in meno
     * (tre sfridi 10, 10 e 6000 danno 84; due sfridi da 100 danno 100). Nella pratica non capita —
     * misurato su 200 distinte casuali, non è successo mai — ma la barra nuova è l'unica cosa che si
     * paga davvero, e non deve dipendere da come cadono i dadi.
     */
    private static boolean meglio(PianoDiTaglio candidato, PianoDiTaglio attuale) {
        if (candidato.barreNuove() != attuale.barreNuove()) {
            return candidato.barreNuove() < attuale.barreNuove();
        }
        return candidato.mediaGeometricaSfrido() < attuale.mediaGeometricaSfrido();
    }
}

File: core/src/main/kotlin/com/cutcalculator/ottimizzatore/MultiStartCasuale.kt
package com.cutcalculator.ottimizzatore

import com.cutcalculator.dominio.Avanzo
import com.cutcalculator.formule.Distinta
import java.util.Random

/**
 * Meta-strategia multi-start, l'euristica **predefinita**: genera più piani candidati e tiene il
 * migliore — meno barre nuove, e a parità la [PianoDiTaglio.mediaGeometricaSfrido] più bassa. È
 * l'uso previsto di quella metrica.
 *
 * Parte da un candidato deterministico (best-fit su ordine decrescente), poi prova `iterazioni`
 * impacchettamenti best-fit con i pezzi in ordine **casuale** (stesso `seed` ⇒ risultato
 * riproducibile). Best-fit su ordine perturbato a volte batte l'ordine decrescente puro: tenendo il
 * migliore non si può fare peggio del baseline.
 *
 * Costa circa **4 volte** il best-fit da solo (misurato su 200 distinte da 20-100 pezzi: 0,7 ms
 * l'una), il che su ordini di questa taglia resta impercettibile. Se un domani le commesse
 * diventassero molto grandi, è la prima impostazione da riportare su "miglior incastro".
 */
open class MultiStartCasuale(private val iterazioni: Int, private val seed: Long) : Ottimizzatore {

    /** Default comodo: 50 tentativi, seed fisso per risultati riproducibili. */
    constructor() : this(50, 0L)

    override fun ottimizza(
        distinta: Distinta, avanzi: List<Avanzo>, lunghezzaBarraStandard: Double
    ): PianoDiTaglio {
        var migliore = Impacchettatore.impacchetta(
            distinta, avanzi, lunghezzaBarraStandard,
            Impacchettatore.DECRESCENTE, Impacchettatore.MIGLIOR_INCASTRO
        )

        val rng = Random(seed)
        repeat(iterazioni) {
            val candidato = Impacchettatore.impacchetta(
                distinta, avanzi, lunghezzaBarraStandard,
                Impacchettatore.casuale(rng), Impacchettatore.MIGLIOR_INCASTRO
            )
            if (meglio(candidato, migliore)) {
                migliore = candidato
            }
        }
        return migliore
    }

    private companion object {
        /**
         * Il confronto fra due piani, in **quest'ordine**: prima chi compra **meno barre nuove**,
         * poi — a parità — chi ha la media geometrica dello sfrido più bassa.
         *
         * La media geometrica da sola non basta come criterio: premia lo sfrido *concentrato*, e un
         * piano con una barra in più può avere una media migliore di uno con una barra in meno (tre
         * sfridi 10, 10 e 6000 danno 84; due sfridi da 100 danno 100). Nella pratica non capita —
         * misurato su 200 distinte casuali, non è successo mai — ma la barra nuova è l'unica cosa
         * che si paga davvero, e non deve dipendere da come cadono i dadi.
         */
        fun meglio(candidato: PianoDiTaglio, attuale: PianoDiTaglio): Boolean {
            if (candidato.barreNuove() != attuale.barreNuove()) {
                return candidato.barreNuove() < attuale.barreNuove()
            }
            return candidato.mediaGeometricaSfrido() < attuale.mediaGeometricaSfrido()
        }
    }
}
